client.go

```diff
@@ -254,76 +290,120 @@ func (c *Client) do(method, urlpath string, doOptions doOptions) (*http.Response
+	defer c.configLock.RUnlock() // This defer matches both the initial and the above lock
 
-	for k, v := range doOptions.headers {
-		req.Header.Set(k, v)
-	}
+	httpClient := c.httpClient
+	endpoint := c.getAPIPath(urlpath, query, doOptions.unversioned)
 
+	// The doOptions Context is shared for every attempted request in the do.
 	ctx := doOptions.context
 	if ctx == nil {
 		ctx = context.Background()
 	}
 
-	resp, err := httpClient.Do(req.WithContext(ctx))
-	if err != nil {
-		// If it is a custom error, return it. It probably knows more than us
-		if serror.IsStorageOSError(err) {
+	var failedAddresses = map[string]struct{}{}
+
+	c.addressLock.Lock()
+	var addresses = make([]string, len(c.addresses))
+	copy(addresses, c.addresses)
+	c.addressLock.Unlock()
+
+	for _, address := range addresses {
+		target := address + endpoint
+
+		req, err := http.NewRequest(method, target, params)
+		if err != nil {
+			// Probably should not try and continue if we're unable
+			// to create the request.
 			return nil, err
 		}
+		req.Header.Set("User-Agent", c.userAgent)
+		if doOptions.data != nil {
+			req.Header.Set("Content-Type", "application/json")
+		} else if method == "POST" {
+			req.Header.Set("Content-Type", "plain/text")
+		}
+		if c.username != "" && c.secret != "" {
+			req.SetBasicAuth(c.username, c.secret)
+		}
 
-		if strings.Contains(err.Error(), "connection refused") {
-			return nil, ErrConnectionRefused
+		for k, v := range doOptions.headers {
+			req.Header.Set(k, v)
 		}
-		return nil, chooseError(ctx, err)
-	}
-	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
-		return nil, newError(resp)
-	}
-	return resp, nil
-}
 
-// if error in context, return that instead of generic http error
-func chooseError(ctx context.Context, err error) error {
-	select {
-	case <-ctx.Done():
-		return ctx.Err()
-	default:
-		return err
+		resp, err := httpClient.Do(req.WithContext(ctx))
+		if err != nil {
+			// If it is a custom error, return it. It probably knows more than us
+			if serror.IsStorageOSError(err) {
+				switch serror.ErrorKind(err) {
+				case serror.APIUncontactable:
+					// If API isn't contactable we should try the next address
+					failedAddresses[address] = struct{}{}
+					continue
+				case serror.InvalidHostConfig:
+					// If invalid host or unknown error, we should report back
+					fallthrough
+				case serror.UnknownError:
+					return nil, err
+				}
+			}
+
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				if _, ok := err.(net.Error); ok {
+					// Be optimistic and try the next endpoint
+					failedAddresses[address] = struct{}{}
+					continue
+				} else {
+					return nil, err
+				}
+			}
+		}
+
+		// If we get to the point of response, we should move any failed
+		// addresses to the back.
+		failed := len(failedAddresses)
+		if failed > 0 {
+			// Copy addresses we think are okay into the head of the list
+			newOrder := make([]string, len(addresses)-failed)
```

Comment:
Like the pre-allocation of the correct size! You could make this a little easier to read (by eliminating the index counter) if you allocate it with a _length_ of 0, but the correct _capacity_ for all the elements:

```go
newOrder := make([]string, 0, len(addresses)-failed)
```

and then instead of having to track `i`:
```go
newOrder[i] = addr
i++
```

you can instead fo `newOrder = append(newOrder, addr)` - you still have the benefit of correct array allocation but less vars to keep in your head!